### PR TITLE
Check confidentiality for whitelabel

### PR DESCRIPTION
**Pull request by sofiyareverse**
https://www.pivotaltracker.com/n/projects/2167210/stories/178387431

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -355,7 +356,7 @@ def token_props(token)
     end
 
     def redirect_for_whitelabel
-      if @whitelabel_mission
+      if @whitelabel_mission && !@project.require_confidentiality?
```
Comment (zhinfuhnya) [inline app/controllers/projects_controller.rb:359]: this condition already applies to `policy(@project).show_contributions?`
so, it's unnecessary double check

look into `app/policies/project_policy.rb`

method `show_contributions?` call `show_contributions_to_everyone?` which check:
`project.public? && !project.require_confidentiality?`

File: app/models/project.rb [hunk 2]
```diff
@@ -80,6 +80,7 @@ class Project < ApplicationRecord
   scope :unarchived, -> { where.not visibility: 4 }
   scope :publics, -> { where 'projects.visibility in(1)' }
   scope :with_all_attached_images, -> { with_attached_image.with_attached_square_image.with_attached_panoramic_image }
+  scope :confidential, -> { where(require_confidentiality: true) }
```
Comment (zhinfuhnya) [inline app/models/project.rb:83]: where does this scope used?

Comment (sofiyareverse) [reply app/models/project.rb:83]: it is a draft. I will continue with it today

File: app/policies/project_policy.rb [hunk 3]
```diff
@@ -18,7 +18,7 @@ def resolve
       if @account
         account.accessable_projects(scope)
       else
-        scope.publics
+        scope.publics.non_confidential
```
Comment (Alexey1100) [inline app/policies/project_policy.rb:21]: Hmm, we should include confidential projects if they are public.
Confidentiality flag just hides transfers/accounts pages, but project overview and tasks/roadmap are still visible.

I guess the issue is that these visible pages are disabled for whitelabel projects, so enabling confidentiality on whitelabel is pretty much equivalent of making the project non-public.

So, maybe we should apply the scope above only for the whitelabel requests.

**Comment by sofiyareverse** [comment]
According to our last conversation I checked the question about request as param in policy scope and have found the next answer: in application_controller method "set_project_scope" we return whitelabels for current domain, so to my mind we shouldn't check request again in the projects_controller method set_project

File: spec/policies/project_policy_spec.rb [hunk 4]
```diff
@@ -58,8 +58,28 @@
...
+        let!(:my_whitelabel_confidential_project) { create(:project, title: 'whitelabel confidential', account: project_account, visibility: 'public_listed', require_confidentiality: true, mission: create(:mission, whitelabel: true)) }
+
+        it 'returns all public non-confidential projects for whitelabels if account is nil' do
+          whitelabel_projects = Project.where(whitelabel: true)
+          projects = ProjectPolicy::Scope.new(nil, whitelabel_projects).resolve
+
+          expect(projects.map(&:title).sort).to eq([my_whitelabel_non_confidential_project].map(&:title).sort)
+        end
+
+        it 'returns all public confidential and non-confidential projects for whitelabels if account is nil' do
```
Comment (Alexey1100) [inline spec/policies/project_policy_spec.rb:76]:
```suggestion
        it 'returns all public confidential and non-confidential projects for whitelabels if account has access' do
```

**Review by Alexey1100 — APPROVED** [review]

**Pull request merged**